### PR TITLE
[scaling] deployer: merged the deployer with gateway state machine

### PR DESCRIPTION
**Pull request by iulianbarbu**
## Description

A draft for a new deployer based on the gateway project state machine and the current deployer. It's a big PR and the best way to go through this is by manual testing of the bigger features for now. Still need to cover functionality around:
* idle projects
* Integration with the builder/logger
* automated tests
* user services reachability (proxying or networking infra that would disallow inter-container communication for projects, but allow the communication that happens through the user network gateway).
* deployer protobuf API for `get_deployments`, `get_services`, `get_service_status`.
* test the user service completed state (as for workers that end up by themselves - it's a bit more special because it's something that updates the states in an aspect-oriented way).

As for nits, there should be significant simplifications coding wise:
* create queries with JOIN for the state query coming from the deployment side (currently we make two queries: (deployment_id -> deployment)-> (service_id -> service -> state).
* certainly others

Outstanding changes:
* unified the state cross deployment/service state machine. The previous deployment state was updated in an aspect-oriented manner, based on the logs gathered by the deployer_layer tracing. Now both the deployment & services state is represented by the service state driven by docker containers state machine.
* clarify if this state unification breaks any assumptions made by dependencies (e.g. console)
* restarting the deployer involves restarting all existing running services to redo the runtime client.

## How has this been tested? (if applicable)
Manually tested the following scenarios:
1) deploy rpc
2) destroy rpc
3) restarting the deployer with running services and properly catching up on their previous state
    3.a) if the service sandbox is still alive and it's state is `Running`, the deployer destroys the service and then it's refreshing it's current state.
    3.b) if the service sandbox is down but its state is `Running`, the deployer re-creates it.

### CODE REVIEW

File: deployer/src/lib.rs [hunk 1]
```diff
@@ -1,4 +1,604 @@
...
+            .serve(bind_address)
+            .await
+            .expect("to serve on address");
+        Ok(())
+    }
+
+    // Ensures this service is created and the runtime loaded & started. Important to note that this method
+    // can be called when starting the deployer, to pick up from persistence the existing deployments and
+    // reinstate them if they are on the running code path, but also when deploying a brand new deployment,
+    // storing it in the persistence.
```
Comment (oddgrd) [inline deployer/src/lib.rs:220]: Super nit: this should be doc comment
```suggestion
    /// Ensures this service is created and the runtime loaded & started. Important to note that this method
    /// can be called when starting the deployer, to pick up from persistence the existing deployments and
    /// reinstate them if they are on the running code path, but also when deploying a brand new deployment,
    /// storing it in the persistence.
```

File: deployer/src/lib.rs [hunk 2]
```diff
@@ -1,4 +1,604 @@
...
+        let req_deployment = match request.deployment {
+            Some(inner) => inner,
+            None => {
+                return Ok(Response::new(DeployResponse {
+                    success: false,
+                    message: Some("invalid argument: missing deployment information".to_string()),
+                    deployment_id: None,
+                }))
+            }
+        };
```
Comment (oddgrd) [inline deployer/src/lib.rs:332]: This and the one below could use let else, e.g. `let Ok(req_deployment) = request.deployment else { return ...}`

File: common/src/builder/Containerfile
Comment (chesedo) [inline common/src/builder/Containerfile:1]: Why is this moved to common?

File: common/src/builder/mod.rs [hunk 3]
```diff
@@ -45,14 +45,20 @@ impl MockedBuilder {
...
         let mut buf = Vec::new();
         f.read_to_end(&mut buf).await?;
-        self.push_image(buf).await.map_err(Error::Oci)?;
-        debug!("Successfuly built and pushed the image to the container registry.");
+
+        // We do not expect multi-arch images to be returned from the builder, so we're expecting
+        // a single image name, corresponding to a single image manifest, for a single architecture.
+        let image_names = self.push_image(buf).await.map_err(Error::Oci)?;
+        if image_names.len() == 1 {
+            todo!();
```
Comment (chesedo) [inline common/src/builder/mod.rs:53]: Is this still a todo?

File: deployer/src/args.rs [hunk 4]
```diff
@@ -1,25 +1,35 @@
...
     /// Address to reach the authentication service at
-    #[clap(long, default_value = "http://127.0.0.1:8008")]
+    #[arg(long, default_value = "http://auth:8008")]
     pub auth_uri: Uri,
 
-    /// Address to bind API to
-    #[clap(long, default_value = "0.0.0.0:8001")]
-    pub api_address: SocketAddr,
+    /// Address to connect to the provisioning service
+    #[clap(long, default_value = "http://provisioner:5001")]
```
Comment (chesedo) [inline deployer/src/args.rs:21]: Some are still `#[clap]`

File: deployer/src/lib.rs [hunk 5]
```diff
@@ -1,4 +1,604 @@
...
+
+        if service.state.is_completed() || service.state.is_destroyed() {
+            return Ok(Response::new(TargetIpResponse {
+                success: false,
+                target_ip: None,
+                message: Some("the service is not running".to_string()),
+            }));
+        }
+
+        if service.state.is_stopped() && request.instate {
```
Comment (oddgrd) [inline deployer/src/lib.rs:496]: Could we simplify this by just returning a target IP or a tonic::Status with a message (rather then the response with success, ip and message)? Then we can use `?`

File: deployer/src/deployment.rs [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+#[instrument(skip(deployment_id, dal), fields(deployment_id = %deployment_id, state = %ServiceCompleted::name()))]
+async fn completed_cleanup<D: Dal + Sync + 'static>(
+    deployment_id: &Ulid,
+    dal: D,
+    // docker: Docker,
+    // runtime_manager: RuntimeManager,
+    // sender: tokio::sync::mpsc::Sender<
+    //     Box<dyn Task<(), Output = (), Error = crate::project::error::Error>>,
+    // >,
+    // task_router: TaskRouter<BoxedTask>,
```
Comment (chesedo) [inline deployer/src/deployment.rs:147]:
```suggestion
```

File: deployer/src/project/docker.rs [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+pub trait ContainerInspectResponseExt {
+    fn container(&self) -> &ContainerInspectResponse;
+
+    fn service_id(&self) -> Result<Ulid, ServiceErrored> {
+        let container = self.container();
+
+        Ulid::from_string(safe_unwrap!(container
+            .config
+            .labels
+            .get("shuttle.service_id")))
```
Comment (oddgrd) [inline deployer/src/project/docker.rs:205]: Should we make these labels consts?

File: deployer/src/deployment.rs [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+        claim: Option<Claim>,
+    ) -> Result<()> {
+        let port = match pick_unused_port() {
+            Some(port) => port,
+            None => {
+                return Err(Error::PrepareRun(
+                    "could not find a free port to deploy service on".to_string(),
+                ))
+            }
+        };
```
Comment (chesedo) [inline deployer/src/deployment.rs:225]: Why are we picking a random port?

File: deployer/src/deployment.rs [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+            );
+
+            error!(%status, "failed to start service");
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    // use std::{
```
Comment (chesedo) [inline deployer/src/deployment.rs:339]: Maybe we should remove these for now?

File: deployer/src/project/service/state/d_started.rs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                let mut last = None;
+
+                while stats.len() > (idle_minutes as usize) {
+                    last = stats.pop_front();
+                }
+
+                if let Some(last) = last {
+                    let cpu_per_minute = (new_stat.cpu_stats.cpu_usage.total_usage
+                        - last.cpu_stats.cpu_usage.total_usage)
+                        / idle_minutes;
```
Comment (chesedo) [inline deployer/src/project/service/state/d_started.rs:123]: This is where we do the idle check everything it is triggered by ambulance / refresh

**Review by oddgrd — COMMENTED** [review]
This is looking good Iulian! I left some questions and comments, mostly nits and refactoring suggestions.

File: deployer/src/project/service/state/f_running.rs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+impl StateVariant for ServiceRunning {
+    fn name() -> String {
+        "Running".to_string()
+    }
+
+    fn as_state_variant(&self) -> String {
+        Self::name()
+    }
```
Comment (chesedo) [inline deployer/src/project/service/state/f_running.rs:27]: Maybe this should be the default impl on the trait definition?

File: deployer/src/runtime_manager.rs
Comment (chesedo) [inline deployer/src/runtime_manager.rs:1]: We should remove the `deployer_alpha` version in this PR too

File: proto/runtime.proto [hunk 12]
```diff
@@ -13,6 +13,9 @@ service Runtime {
   // Stop a started service
   rpc Stop(StopRequest) returns (StopResponse);
 
+  // Ping a service to check the health
+  rpc HealthCheck(Ping) returns (Pong);
```
Comment (chesedo) [inline proto/runtime.proto:17]: Nice :partying_face: :tada:

**Review by chesedo — APPROVED** [review]
Left a few small comments, but we can merge directly after.

I kept notes in C24 for bigger things we can test
